src/linker/Linker.Steps/RemoveUnreachableBlocksStep.cs

```diff
@@ -102,53 +87,274 @@ void RewriteBodies (Collection<TypeDefinition> types)
+					// Loop was detected - the stack hasn't changed since the last time we tried to process this method
+					// as such there's no way to resolve the situation (running the code below would produce the exact same result).
+
+					// Observation:
+					//   All nodes on the stack which have `LastAttemptStackVersion` equal to `processingStackVersion` are part of the loop
+					//   meaning removing any of them should break the loop and allow to make progress.
+					//   There might be other methods in between these which don't have the current version but are dependencies of some of the method
+					//   in the loop.
+					//   If we don't process these, then we might miss constants and branches to remove. See the doc
+					//   `constant-propagation-and-branch-removal.md` in this repo for more details and a sample.
+
+					// To fix this go over the stack and find the "oldest" node with the current version - the "oldest" node which
+					// is part of the loop:
+					var lastNodeWithCurrentVersion = stackNode;
+					for (var currentNode = stackNode; currentNode != null; currentNode = currentNode.Next) {
+						if (currentNode.Value.LastAttemptStackVersion == processingStackVersion)
+							lastNodeWithCurrentVersion = currentNode;
+					}
+
+					// There should be at least 2 nodes with the latest version to form a loop
+					Debug.Assert (lastNodeWithCurrentVersion != stackNode);
+
+					// Now go back over all nodes from the "oldest" one back to the top and find any nodes which are not of current version.
+					// For all of them, move them to the top of the stack.
+					var candidateNodeToMoveToTop = lastNodeWithCurrentVersion;
+					bool foundNodesWithNonCurrentVersion = false;
+					while (candidateNodeToMoveToTop != stackNode) {
+						var previousNode = candidateNodeToMoveToTop.Previous;
+						if (candidateNodeToMoveToTop.Value.LastAttemptStackVersion != processingStackVersion) {
+							processingStack.Remove (candidateNodeToMoveToTop);
+							processingStack.AddFirst (candidateNodeToMoveToTop);
+							foundNodesWithNonCurrentVersion = true;
+						}
+
+						candidateNodeToMoveToTop = previousNode;
+					}
+
+					// If any node was found which was not of current version (and moved to the top of the stack), move on to processing
+					// the stack - this will give a chance for these methods to be processed. It doesn't break the loop and we should come back here
+					// again due to the same loop as before, but now with more nodes processed (hopefully all of the dependencies of the nodes in the loop).
+					// In the worst case all of those nodes will become part of the loop - in which case we will move on to break the loop anyway.
+					if (foundNodesWithNonCurrentVersion) {
+						continue;
+					}
+
+					// No such node was found -> we only have nodes in the loop now, so we have to break the loop.
+					// We do this by processing it with special flag which will make it ignore any unprocessed dependencies
```

Comment:
The comment suggests a better name for the field, which part of the name the bool value `treatUnprocessedAsNonConst` does change?

Reply:
Sorry - I don't think I follow...
I guess the name could mention that it's for breaking loops - but it didn't felt right to me. We are using it to break loops, sure, but what it does is in its name - it will treat all unprocessed dependencies as non-const.

I'll change it to `treatUnprocessedDependenciesAsNonConst`.

Reply:
Let me ask different, what does `treatUnprocessedDependenciesAsNonConst = false` do ?

Reply:
If there's at least one unprocessed dependency, the processing of the current method will be stopped and we go back to the stack - in the code/comments I call this "backing off". The method will be retried at some point.